File: python/mean_median_mode.py
import sys

print("Welcome to mean_median_mode!\n")

def find_mean(nums): # add option for decimals
    sum = 0
    decimal_places = int(input("How many decimal places would you like to round to? >> "))
    
    for num in nums:
        sum += num
    
    if decimal_places == 0:
        return int(sum / len(nums))
    else: 
        return round(sum / len(nums), decimal_places)

def find_median(nums):
    nums.sort() 
    if len(nums) % 2 == 1: 
        return nums[int((len(nums) - 1) / 2)]
    else:
        return nums[0:int(len(nums) / 2)][-1], nums[int(len(nums) / 2):][0]
        
def find_mode(nums):
    current_highest = 1
    result = []
    for num in nums:
        if nums.count(num) >= current_highest and num not in result: 
            result.append(num)
            current_highest = nums.count(num)
    return result
                       
nums_str = input("\nEnter numbers seperated by commas (ex. 1, 2, 33, 4) >> ")
nums = [int(num) for num in nums_str.split(',')]

print("\nMean = {}\n".format(str(find_mean(nums))))
print("Median = {}\n".format(str(find_median(nums))))
print("Mode = {}\n".format(str(find_mode(nums))))

File: java/MeanMedianMode.java
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class MeanMedianMode {
  public static void main(String[] args) {
    List<Integer> nums = getNums();

    String mean = getMean(nums);
    List<Integer> median = getMedian(nums);
    List<Integer> mode = getMode(nums);

    System.out.println("Mean: " + mean);
    System.out.println("Median: "
        + median.toString()
          .substring(1, median.toString()
          .length() - 1));
    System.out.println("Mode: " + mode.toString().substring(1, mode.toString().length() - 1));

  }

  private static List<Integer> getNums() {
    String numStr = getInput("Enter a list of numbers seperated by commas (ex. 1, 2, 3): ");
    return Stream.of(numStr.split(","))
        .map(String::trim)
        .map(Integer::parseInt)
        .collect(Collectors.toList());
  }

  private static String getInput(String msg) {
    System.out.println(msg);
    Scanner s = new Scanner(System.in);
    String input = s.nextLine();
    return input;
  }

  public static String roundMean(Double num) {
    int decimalPlaces = Integer.parseInt(getInput("Enter number of decimal places to round to: "));
    BigDecimal bd = new BigDecimal(Double.toString(num));
    bd = bd.setScale(decimalPlaces, RoundingMode.HALF_UP);
    return Double.toString(bd.doubleValue());
  }

  public static String getMean(List<Integer> nums) {
    double mean = (double) nums.stream().mapToInt(Integer::intValue).sum() / nums.size();
    return roundMean(mean);
  }

  public static List<Integer> getMedian(List<Integer> nums) {
    Collections.sort(nums);
    List<Integer> median = new ArrayList<Integer>();
    median.add(nums.get(nums.size() / 2));
    if (nums.size() % 2 == 0) {
      median.add(nums.get((nums.size() / 2) - 1));
    }

    return median;
  }

  public static List<Integer> getMode(List<Integer> nums) {
    Map<Integer, Long> counts = nums.stream()
        .collect(Collectors.groupingBy(e -> e, Collectors.counting()));
    Long max = counts.entrySet().stream()
          .max((a, b) -> a.getValue() > b.getValue() ? 1 : -1)
          .get().getValue();

    return counts.entrySet().stream()
        .filter(e -> e.getValue() == max)
        .map(e -> e .getKey())
        .collect(Collectors.toList());
  }
}
